### ISSUE TITLE
DSP Precision Question

### ISSUE THREAD

**Issue by XYC-EXP** [sadko4u/lsp-plugins#267]
I have recently tested the precision of the lsp parametric equaliser. I enabled a couple of filters and set their gain to 0 db. The test signal was white noise. Then I noticed that the output after the equaliser changed a little compared to the original signal. And the SNR did not reach 138 dB. 

When using fft mode, the error rms is about -111dB, and eq10q is about -114dB.
The error is also a white noise.

Is this because the lsp plugin doesn't use high precision processing or achieve 24bit processing precision is hard? What strikes me as odd is that the eq10q is not as high precision as desired either.

Is there a way to solve this problem?

**Comment by sadko4u** [comment]
LSP Plugins use IEEE 754 32-bit floating point number when processing the signal.
But for better performance LSP Plugins enable the DAZ/FTZ CPU feature that interprets denormal floating-point values as zeros because denormal values issue serious CPU penalty in computation.
Moreover, dependent on the CPU architecture and CPU instruction set, it may utilize different SIMD instructions which may yield to the different precision of the result. For example, the FMA operations are declared being more precise rather than simultaneous multiplication and addition instructions but dependent on your CPU's instruction set support they can be turned on (if supported) or off (if not).
LSP Plugins don't try to reproduce any other equalizers like eq10q.
There are many commercial and non-commercial equalizers that tend to introduce additional noise and distortion to the signal, but people are happy to use them. So I don't think that it is a serious problem.

**Comment by XYC-EXP** [comment]
Yes. It's true that float32 will produce pretty good sound. However, some effects may occur when errors are accumulated. For example, when using iir, the music sounds more powerful but there are some loss of dynamics.(maybe it's the low frequency part, I haven't tested it yet.) When using fft, the music sounds bouncy and warmer.

I have make another two test: generate a uniform white noise(range from -1.0 to 1.0), do fft and then do ifft.

float32(float) fft: SNR: 133.26 dB (< 24 bit quantization) (greater than some hardware)
                           max bit change: won't exceed 20bit quantization in most cases
float64(double) fft: SNR: 304.92 dB
                           max bit change: won't exceed 49bit quantization in most cases

So I would suggest that, for operations where errors will accumulate many times(adjust audio volume < fft < iir < full convolution), using float64(double) would be best if a ultra high quality music is needed. 

For example, limiter or compressor don't require float64(double), because the error in RMS calculation won't affect sound quality so much. Equalizer and Multiband-Compressor includes filtering so that they need to introduce float64(double) to achieve ultra high quality.

In my vision, LSP plugin will have a button to enable the high precision processing mode. In this mode, the plugin will only execute slower and unoptimised high precision code. (this mode is only designed to export the final product, speed is not particularly important, therefore, does not need to be optimised if there isn't much time for this.)

😃

**Comment by sadko4u** [comment]
Even so, this won't be introduced in the nearest future.
That requires implementation of many things inside of the bundle.
Additionally, some hardware platforms still don't support SIMD optimizations of double precision floating point numbers.

**Comment by XYC-EXP** [comment]
OK. Thank you.
